File: PA23.java
Time complexity : O(n^2)
Space Complexity: O(1) 

class Solution {
    public int minFallingPathSum(int[][] A) {
       
        int msum=Integer.MAX_VALUE;
        if(A.length ==1){
            for(int j=0;j<A[0].length;j++){
                msum = Math.min(msum, A[0][j]);
            }
            return msum;
        }
        
        for(int i=1;i<A.length;i++){
           
            for(int j=0;j<A[0].length;j++){

                if(j==0 || j== A[0].length-1 ){
                    if(j==0){
                         A[i][j]= A[i][j] + Math.min(A[i-1][j],A[i-1][j+1]);
                    }
                    
                    else{
                        A[i][j]= A[i][j] + Math.min(A[i-1][j],A[i-1][j-1]);
                    }
                    
                }
                else{
                    A[i][j]= A[i][j] + Math.min(A[i-1][j-1],Math.min(A[i-1][j],A[i-1][j+1]));
                }
                
                    if(i==A.length-1){
                        msum= Math.min(msum,A[i][j]);
                    }
            }
        }
        return msum;
    }
}

File: PA_23.py
#LC 931 - Minimum Falling Path Sum
#Time Complexity - O(n^2)
#Space Complexity - O(1)
class Solution(object):
    def minFallingPathSum(self, A):
        """
        :type A: List[List[int]]
        :rtype: int
        """
   
        for i in range(1,len(A)):
            for j in range(len(A[0])):
                if j == 0:
                    A[i][j] = A[i][j] + min(A[i-1][j],A[i-1][j+1])
                elif j == len(A[0])-1:
                    A[i][j] = A[i][j] + min(A[i-1][j],A[i-1][j-1])
                else:
                    A[i][j] = A[i][j] + min(A[i-1][j],A[i-1][j-1],A[i-1][j+1])
        
        
        return min(A[-1])